docs/imds_retry_based_on_errors.md

```diff
@@ -0,0 +1,65 @@
+# MSAL & MSI IMDS Error Handling and Retry Strategy Specification
+
+## Overview
+This document defines the error handling and retry strategy for MSAL when interacting with the IMDS (Instance Metadata Service) endpoint for Managed Identity (MSI) token acquisition.
+
+---
+
+## 1️⃣ HTTP Status Codes & Recommended Actions
+
+| **HTTP Status Code** | **Error Reason**                               | **Recommended Action**                      | **Retry Delay Strategy**                 |
+|----------------------|-----------------------------------------------|---------------------------------------------|-----------------------------------------|
+| **404**             | IMDS endpoint is updating                     | Retry with Exponential Backoff              | **1s → 2s → 4s**                        |
+| **410**             | IMDS is undergoing updates                    | Wait up to **70 seconds**, then retry      | **70s (fixed wait)**                    |
+| **429**             | IMDS Throttle limit reached                    | Retry with Exponential Backoff              | **2s → 4s → 8s → 16s**                  |
+| **408**             | Request Timeout                                | Retry with Exponential Backoff              | **1s → 2s → 4s**                        |
+| **5xx**             | Transient service error                        | Retry with Exponential Backoff **(max 3 retries)** | **1s → 2s → 4s**             |
+| **400**             | Bad Request (Invalid Parameters)               | **Do not retry**, fix request               | **No retry**                             |
+| **401**             | Unauthorized                                  | **Do not retry**, check authentication setup | **No retry**                             |
+| **403**             | Forbidden                                     | **Do not retry**, verify permissions       | **No retry**                             |
+| **timeout**         | IMDS endpoint is updating                     | Retry with Exponential Backoff              | **1s → 2s → 4s**                        |
+
+---
+
+## 2️⃣ Identity Propagation & Special Handling for "Identity Not Found" Errors
+- **Scenario:** When an identity is newly assigned to a VM, it may take time for the IMDS service to recognize the identity.
+- **Exception Handling:**  
+  - If the **IMDS response contains "Identity Not Found"**, retry the request using **exponential backoff**.
```

Comment:
What error code(s) will return "Identity Not Found"?

Reply:
updated